212 - Document LOQ, Review and add Workflow Diagrams (#628)
openmsupply/application-manager-web-app: documentation/internal/Level-1-Review-And-LOQ.md

# Level 1 Review and LOQ

The aim of this page is to outline the structure and flow of:

- Reviewer - reviewing an application
- Reviewer - submitting LOQ (list of questions or change request)
- Applicant - making changes in response to LOQ
- Applicant - re-submitting application with changes
- Reviewer - re-reviewing changes that were made

In the context of this document, we are assuming that reviewer is already assigned, all sections and they are the only level reviewer defined for a given stage.

## Brief Summary

Below is the summary of logic aspects of this topic

### Back End

> - Changes status of review and application to inform front end of actions that can be made

Reviewer: ```suggestion
- Changes `review.status` and `application.status` to inform front end of actions that can be made
```

> - Trims Responses that are duplicated and not changed

Reviewer: ```suggestion
- On review **submission** trim `review_responses` that have been created but not reviewed by user
- On application **re-submission**, trim `application_response` that are duplicated and not changed
- On review **re-submission**, trim `review_responses` that are duplicated and not changed
```

### Front End

- Calculates actions and progress of review and application
- GraphQL mutation to create review and application responses

> - GraphQL mutation to update review_responses

Reviewer: ```suggestion
- GraphQL mutation to update `application_response` or `review_response`
```

- Validates submission of review and application
- GraphQL mutation to request submission of review and application

### Structure

> The 'state' of application and review (actions, progress, etc..) is calculated and stored in one state structure. It aids visual component and de-couples logic from UI elements. This structure is recalculated whenever a response is changed

Reviewer: That's very good, I just missed a little more explanation around what hooks runs in what case, would we require a little more around that? For example, does the `useGetApplicationStructure` re-runs when there is an update to `review_response`?

Author: I've added a little bit more: d887cf212f7301f79931fee9c9a53db6409fbc3b, didn't quite add about review_response only affecting only useGetReviewStructureForSection only.  Was hoping that diagram would describe it better:

![Screen Shot 2021-05-21 at 12 22 08 AM](https://user-images.githubusercontent.com/26194949/118977747-9dfbf680-b9ca-11eb-8c46-22836fc7763b.png)

## Schema

The aim of this diagram is to give some context (an anchor point), it's an abstracted schema diagram, not all of the tables involved in Review and LOQ process are shown

![Review and LOQ Schema](images/Review-And-LOQ-Schema.png)

> This schema structure allows new responses to be created (previous responses to be duplicated) for every new update, and this creates a system which maintains historic data integrity.

Reviewer: ```suggestion
This schema structure allows new responses to be created (previous responses to be duplicated) for every change on the application status, and this creates a system which maintains historic data integrity.
```

## Starting and Re-Staring Review

**Front End**

> When reviewer start/re-starts a review, new review_responses are created for every `latest` application_response. If latest application_response already has a review_response linked (from previous review), review_response is duplicated. `useRestartReview` and `useCreateReview` hooks.

Reviewer: ```suggestion
When reviewer start/re-starts a review, new `review_response` entities are created for every `latest` `application_response`. If latest `application_response` already has a `review_response` linked (from previous review), `review_response` is duplicated. Done by `useRestartReview` and `useCreateReview` hooks.
```

> `Latest` application response is just application response with latest time_updated for a given template_element.

Reviewer: ```suggestion
`latest_application_response` is just application response with latest `time_updated` for a given element.
```

It's possible that latest application response is not visible (it could become invisible due to changes in application), these are ignored.

> We only create review response for template_elements that are assigned to reviewer via review_question_assignment (see [Assignment Process](Assignment-Process.md) for more details)

Reviewer: ```suggestion
We only create `review_response` for elements that are assigned to reviewer via `review_question_assignment` (see [Assignment Process](Assignment-Process.md) for more details)
```

**Back End**

> When review is submitted, an action will `trim` any responses that haven't changed and will update latest review responses time_updated to match review submission time (latest from status and stage history table) `trimResponses` action.

Reviewer: ```suggestion
When review is submitted, an action will `trim` any responses that haven't changed and will update latest review responses `time_updated` to match review submission time (latest from status and stage history table) `trimResponses` action.
```

**Review Status**

We will be restricting editing of review tables (apart from `trigger`) via postgres row level policies. Front end requests status change by setting `trigger` on review table to `onReviewCreate`, `onReviewSubmit`, `onReviewRestart`, and back end changes this via `changeStatus` action. Front end needs to wait for `trigger` to reset to `null` before displaying new state of review

## Re-Staring Application

In response to LOQ, an applicant can restart an application. Similar logic to the one described in above section applies.

## Review Decision

> Is created for each iteration of a review, can be of status: `'List of questions`, `Conform`, `Non-conform`, `No Decision`. See `Validation of Submission` below for choices that will be available to the user.

Reviewer: ```suggestion
Is created for each iteration of a review, can be of `status`: 
- `List of questions`
- `Conform`
- `Non-conform`
- Default: `No Decision`

 See `Validation of Submission` below for choices that will be available to the user.
```

## Review Responses Visibility

> `is_visible_to_applicant` field (not editable by any user, only by back end) controls visibility of review_response from applicant perspective. And `recommended_applicant_visibility` field (which is editable by reviewer) suggests visibility of review_response to applicant. `updateReviewVisibility` back end action will set `is_visible_to_applicant` when review is submitted. See `LOQ Workflow Diagram` below

Reviewer: ```suggestion
`is_visible_to_applicant` field (not editable by any user, only by back end) controls visibility of `review_response` from applicant perspective. And `recommended_applicant_visibility` field (which is editable by reviewer) suggests visibility of review_response to applicant. `updateReviewVisibility` back end action will set `is_visible_to_applicant` when review is submitted. See `LOQ Workflow Diagram` below
```

## Actions

Front end determines user actions based on review and application status. And theses statuses are changed by back end via actions, here is an abstracted diagram:

![Review-And-LOQ-Actions](images/Review-And-LOQ-Actions.png)

Actions for review are computed by section via `useGetReviewStructureForSections` -> `generateReviewStructure` -> `generateReviewSectionActions`

Actions for applications `changeRequestsProgress` and `isChangeRequest` is used to display actions in Application Home pages, those are generated in `useGetApplicationStructure` -> `addApplicantChangeRequestStatusToElement` and `generateApplicantChangesRequestedProgress`.

See `Structure` below for more details

## Validation of Submission

When **application** is restarted as a result of LOQ, it can only be submitted if general validation passes AND all questions with change requests have been changed. Logic for this is:

> latestApplicationResponse is different to previousApplicationResponse where previousApplicationResponse is linked to non confirm reviewResponse

Reviewer: ```suggestion
`latestApplicationResponse` is different to `previousApplicationResponse` where `previousApplicationResponse` is linked to non confirm `reviewResponse`
```

**review** can be submitted as `LOQ` or `Non-conform` when at least review_response is marked as `decline`. Otherwise submission of `Conform` can be made if ALL assigned visible latest application responses have `approve` status review responses.

One thing to note, since we always look at latest application response and it's linked review response, we guarantee that all new changes by applicant will be re-reviewed

## Structure

Application and Review Structure are generated in hooks `useGetApplicationStructure` and `useGetReviewStructureForSection` respectively, these structures are re-computed whenever responses are changed. In the context of Review and LOQ, these fields are added:

### Application

![Applicant LOQ Structure Changes](images/Applicant-LOQ-Structure-Changes.png)

### Review

![Level-1-Review-Structure-Changes](images/Level-1-Review-Structure-Changes.png)

## Components and Hooks

> This is described in Application and Review Flow documentation

Reviewer: Link?

Author: 911facfd91e4148c0054bdaf5e8d4a5e50d38966

# LOQ Workflow Diagram

> ![LOQ Workflow](images/LOQ-Workflow.png)

Reviewer: Is this related to `recommend_applicant-visibility`?
![Screen Shot 2021-04-22 at 12 22 43 PM](https://user-images.githubusercontent.com/16461988/115637956-839c1200-a365-11eb-9937-14f70c50360a.png)

Also, the bold is half-way through in the first `Visibility field` part of this diagram:
![Screen Shot 2021-04-22 at 12 23 56 PM](https://user-images.githubusercontent.com/16461988/115638015-a1697700-a365-11eb-89a9-4083d206dc8b.png)

Author: > Is this related to recommend_applicant-visibility?
yep: 6c760856c386cb9425471d64fbdca35dd048eb99
> Also, the bold is half-way through in the first Visibility field part of this diagram:
6c760856c386cb9425471d64fbdca35dd048eb99

# Responses Workflow Diagram

Note this diagram may be slightly outdated, but it is still indicative of implemented logic. It also goes beyond Review and LOQ topic

> ![Responses Workflow](images/Responses-Workflow.png)

Reviewer: Haven't done a review for this one. Will leave it for re-review, but have in mind it doesn't need to be exact up to date, just reflect what we have been doing currently.

Author: I think it's generally pretty accurate. It should give a good idea of the implementation. I wish there was an easier way to update diagram and even link them to code they relate to
